benchmarks/nemo/nemolite2d/manual_versions/psykal_c_opencl/time_step.c

```diff
@@ -232,108 +259,133 @@ void c_invoke_time_step(
         check_status("clCreateBuffer", ret);
         va_device = clCreateBuffer(context, CL_MEM_READ_WRITE, buff_size,
 			                       NULL, &ret);
+        *va_dp = &va_device;
         num_buffers++;
         check_status("clCreateBuffer", ret);
 
         /* Mesh scale factors */
         area_t_device = clCreateBuffer(context, CL_MEM_READ_ONLY, buff_size,
 			                        NULL, &ret);
+        *area_t_dp = &area_t_device;
         num_buffers++;
         check_status("clCreateBuffer", ret);
         area_u_device = clCreateBuffer(context, CL_MEM_READ_ONLY, buff_size,
 			                        NULL, &ret);
+        *area_u_dp = &area_u_device;
         num_buffers++;
         check_status("clCreateBuffer", ret);
         area_v_device = clCreateBuffer(context, CL_MEM_READ_ONLY, buff_size,
 			                        NULL, &ret);
+        *area_v_dp = &area_v_device;
         num_buffers++;
         check_status("clCreateBuffer", ret);
         dx_u_device = clCreateBuffer(context, CL_MEM_READ_ONLY, buff_size,
 			                        NULL, &ret);
+        *dx_u_dp = &dx_u_device;
         num_buffers++;
         check_status("clCreateBuffer", ret);
         dx_v_device = clCreateBuffer(context, CL_MEM_READ_ONLY, buff_size,
 			                        NULL, &ret);
+        *dx_v_dp = &dx_v_device;
         num_buffers++;
         check_status("clCreateBuffer", ret);
         dx_t_device = clCreateBuffer(context, CL_MEM_READ_ONLY, buff_size,
 			                        NULL, &ret);
+        *dx_t_dp = &dx_t_device;
         num_buffers++;
         check_status("clCreateBuffer", ret);
         dy_u_device = clCreateBuffer(context, CL_MEM_READ_ONLY, buff_size,
                                      NULL, &ret);
+        *dy_u_dp = &dy_u_device;
         num_buffers++;
         check_status("clCreateBuffer", ret);
         dy_v_device = clCreateBuffer(context, CL_MEM_READ_ONLY, buff_size,
                                      NULL, &ret);
+        *dy_v_dp = &dy_v_device;
         num_buffers++;
         check_status("clCreateBuffer", ret);
         dy_t_device = clCreateBuffer(context, CL_MEM_READ_ONLY, buff_size,
                                      NULL, &ret);
+        *dy_t_dp = &dy_t_device;
         num_buffers++;
         check_status("clCreateBuffer", ret);
         tmask_device = clCreateBuffer(context, CL_MEM_READ_ONLY,
 				        (size_t)(total_size*sizeof(cl_int)), NULL, &ret);
+        *tmask_dp = &tmask_device;
         num_buffers++;
         check_status("clCreateBuffer", ret);
 
         /* Coriolis parameters */
         gphiu_device = clCreateBuffer(context, CL_MEM_READ_ONLY, buff_size,
 				                      NULL, &ret);
+        *gphiu_dp = &gphiu_device;
         num_buffers++;
         check_status("clCreateBuffer", ret);
         gphiv_device = clCreateBuffer(context, CL_MEM_READ_ONLY, buff_size,
 				                      NULL, &ret);
+        *gphiv_dp = &gphiv_device;
```

Comment:
Well, I was able to remove the buffer device objects that I had as global variables, but I still create a local buffer that I store in a heap allocated cl_mem object to prevent an automatic release when device_buffer goes out of scope.